File: tests/mission_workflow_v1.js
const assert = require('assert');
const fs = require('fs');
const vm = require('vm');
const executionGuidance = require('../public/execution_guidance.js');
const contractGuidance = require('../public/contract_guidance.js');
const cardData = require('../public/card_data.js');

const html = fs.readFileSync('public/index.html', 'utf8');
const inline = [...html.matchAll(/<script(?![^>]*src=)[^>]*>([\s\S]*?)<\/script>/gi)][0][1];

function elementStub(id = '') {
  return {
    id,
    innerHTML: '',
    textContent: '',
    value: '',
    checked: false,
    dataset: {},
    style: {},
    selectedOptions: [{ textContent: '' }],
    classList: { add() {}, remove() {}, toggle() {} },
    addEventListener() {},
    removeEventListener() {},
    setAttribute() {},
    appendChild() {},
    remove() {},
    scrollIntoView() {},
  };
}

const elements = new Map();
function getElement(id) {
  if (!elements.has(id)) elements.set(id, elementStub(id));
  return elements.get(id);
}

const storage = {};
const context = {
  console,
  Date,
  Math,
  Number,
  String,
  Boolean,
  Array,
  Object,
  JSON,
  RegExp,
  performance: { now: () => 100 },
  URLSearchParams,
  encodeURIComponent,
  decodeURIComponent,
  setTimeout: () => 1,
  clearTimeout: () => {},
  setInterval: () => 1,
  clearInterval: () => {},
  alert: () => {},
  confirm: () => false,
  fetch: () => Promise.reject(new Error('network disabled in mission workflow test')),
  localStorage: {
    getItem: key => storage[key] || null,
    setItem: (key, value) => { storage[key] = String(value); },
    removeItem: key => { delete storage[key]; },
  },
  sessionStorage: {
    getItem: key => storage[`session:${key}`] || null,
    setItem: (key, value) => { storage[`session:${key}`] = String(value); },
    removeItem: key => { delete storage[`session:${key}`]; },
  },
  document: {
    body: { appendChild() {} },
    getElementById: getElement,
    querySelector: () => null,
    querySelectorAll: () => [],
    createElement: () => elementStub(),
    addEventListener() {},
    removeEventListener() {},
  },
  navigator: { clipboard: { writeText: () => Promise.resolve() } },
};

context.window = {
  KairosExecutionGuidance: executionGuidance,
  KairosContractGuidance: contractGuidance,
  KairosCardData: cardData,
  open: () => {},
  addEventListener: () => {},
  removeEventListener: () => {},
  matchMedia: () => ({ matches: false, addEventListener() {}, removeEventListener() {} }),
};
context.self = context.window;
context.globalThis = context;

vm.createContext(context);
vm.runInContext(inline, context);
vm.runInContext('latestScannerMeta = { stock_mission_workflow_v1: true, stock_event_memory_presentation_v1: false };', context);

function setup(overrides = {}) {
  return {
    ticker: 'AAA',
    timeframe: '4H',
    direction: 'LONG',
    price: 100,
    entry: 99,
    sl: 95,
    tp1: 108,
    tp2: 112,
    tp3: 118,
    setupGrade: 'A',
    signal_timestamp: '2026-08-05T14:00:00Z',
    ranking_status_bucket: 'ALMOST_READY',
    normalized_status_bucket: 'ALMOST_READY',
    ranking: {
      rank: 1,
      tier: 'TOP_OPPORTUNITY',
      score: 95,
      status_bucket: 'ALMOST_READY',
      priority_bucket: 1,
    },
    ...overrides,
  };
}

const almostReady = setup({ ticker: 'AR', ranking: { rank: 2, tier: 'HIGH_PRIORITY', score: 90, status_bucket: 'ALMOST_READY' } });
const earlyEntry = setup({ ticker: 'EE', ranking_status_bucket: 'EARLY_ENTRY', normalized_status_bucket: 'EARLY_ENTRY', ranking: { rank: 3, tier: 'HIGH_PRIORITY', score: 88, status_bucket: 'EARLY_ENTRY' }, execution_lifecycle_state: 'EARLY_ENTRY_BUILDING' });
const earlyTouch = setup({ ticker: 'ET', ranking_status_bucket: 'EARLY_ENTRY', normalized_status_bucket: 'EARLY_ENTRY', ranking: { rank: 4, tier: 'HIGH_PRIORITY', score: 86, status_bucket: 'EARLY_ENTRY' }, execution_lifecycle_state: 'EARLY_TOUCH' });
const waitingRetest = setup({ ticker: 'RT', execution_lifecycle_state: 'WAITING_FOR_RETEST', ranking: { rank: 5, tier: 'HIGH_PRIORITY', score: 84, status_bucket: 'ALMOST_READY' } });
const tradeNow = setup({ ticker: 'NOW', ranking_status_bucket: 'ENTER_NOW', normalized_status_bucket: 'ENTER_NOW', ranking: { rank: 1, tier: 'TOP_OPPORTUNITY', score: 97, status_bucket: 'ENTER_NOW' }, execution_lifecycle_state: 'ENTRY_TRIGGERED' });
const resolved = setup({ ticker: 'MISS', ranking_status_bucket: 'EARLY_ENTRY', normalized_status_bucket: 'EARLY_ENTRY', ranking: { rank: 6, tier: 'HIGH_PRIORITY', score: 82, status_bucket: 'EARLY_ENTRY' }, execution_lifecycle_state: 'MISSED_ENTRY' });

assert.strictEqual(context.stockMissionWorkflowEnabled(), true);
assert.strictEqual(context.stockEventMemoryPresentationEnabled(), true, 'mission flag activates lifecycle presentation labels');
assert.strictEqual(context.missionWorkflowBucket(almostReady), 'WATCH_CLOSELY');
assert.strictEqual(context.missionWorkflowBucket(earlyEntry), 'WATCH_CLOSELY');
assert.strictEqual(context.missionWorkflowBucket(earlyTouch), 'WATCH_CLOSELY');
assert.strictEqual(context.missionWorkflowBucket(waitingRetest), 'WATCH_CLOSELY');
assert.strictEqual(context.missionWorkflowBucket(tradeNow), 'TRADE_NOW');
assert.strictEqual(context.missionWorkflowBucket(resolved), 'RESOLVED');
const samePlanEarly = setup({ ticker: 'SAME', execution_lifecycle_state: 'EARLY_TOUCH' });
const samePlanRetest = setup({ ticker: 'SAME', execution_lifecycle_state: 'WAITING_FOR_RETEST' });
assert.strictEqual(context.missionSetupIdentity(samePlanEarly), context.missionSetupIdentity(samePlanRetest), 'same setup identity survives lifecycle-only changes');
const replacedPlan = setup({ ticker: 'SAME', entry: 101, execution_lifecycle_state: 'EARLY_TOUCH' });
assert.notStrictEqual(context.missionSetupIdentity(samePlanEarly), context.missionSetupIdentity(replacedPlan), 'material plan change gets a new identity');

const rows = [almostReady, earlyEntry, earlyTouch, waitingRetest, tradeNow, resolved, tradeNow];
const rankingBefore = JSON.stringify(rows.map(row => row.ranking));
const buckets = context.missionRows(rows);
assert.strictEqual(buckets.tradeNow.length, 1, 'no duplicate card across Trade Now');
assert.strictEqual(buckets.watchClosely.length, 4, 'resolved setup leaves Watch Closely');
assert.strictEqual(buckets.watchClosely[0].ticker, 'RT', 'Waiting for Retest is top Watch Closely priority');
assert.strictEqual(JSON.stringify(rows.map(row => row.ranking)), rankingBefore, 'mission helpers do not mutate frozen ranking');

context.renderMissionWorkflow(rows);
const htmlOut = getElement('missionWorkflow').innerHTML;
assert.ok(htmlOut.includes('Trade Now'));
assert.ok(htmlOut.includes('No fully confirmed trades right now') === false);
assert.ok(htmlOut.includes('Watch Closely'));
assert.ok(htmlOut.includes('WAITING FOR RETEST'));
assert.ok(!htmlOut.includes('MISS</span>'), 'resolved setup is not rendered as a current opportunity');

vm.runInContext('latestScannerMeta = { stock_mission_workflow_v1: false };', context);
context.renderMissionWorkflow(rows);
assert.strictEqual(getElement('missionWorkflow').style.display, 'none');

console.log('mission_workflow_v1.js passed');

File: tests/mission_workflow_v1.py
#!/usr/bin/env python3
"""Regression tests for mission workflow presentation metadata."""

from __future__ import annotations

import copy
import sys
from pathlib import Path

ROOT = Path(__file__).resolve().parents[1]
if str(ROOT) not in sys.path:
    sys.path.insert(0, str(ROOT))

import scanner  # noqa: E402


def setup(
    ticker: str,
    rank: int,
    ranking_bucket: str,
    grade: str = "A",
    *,
    lifecycle_state: str | None = None,
    tier: str = "TOP_OPPORTUNITY",
    entry: float = 100.0,
    stop: float = 95.0,
    tp1: float = 110.0,
) -> dict:
    row = {
        "ticker": ticker,
        "timeframe": "4H",
        "direction": "LONG",
        "entry": entry,
        "sl": stop,
        "tp1": tp1,
        "setupGrade": grade,
        "signal_timestamp": "2026-08-05T14:00:00Z",
        "entryStatus": "Tradeable",
        "ranking_status_bucket": ranking_bucket,
        "ranking": {
            "rank": rank,
            "tier": tier,
            "score": 90 - rank,
            "status_bucket": ranking_bucket,
            "priority_bucket": rank,
        },
    }
    if lifecycle_state:
        row["early_entry_shadow"] = {"state": lifecycle_state}
        row["execution_lifecycle_state"] = lifecycle_state
    if ranking_bucket == "ENTER_NOW" and lifecycle_state == "ENTRY_TRIGGERED":
        row["new_entry_signal"] = {
            "bucket": "ENTER_NOW",
            "current_strategy_status": "ENTER_NOW",
            "lifecycle_state": "ENTRY_TRIGGERED",
            "actionable": True,
        }
    return row


def test_bucket_classification_and_strategy_fields_are_read_only():
    rows = [
        setup("NOW", 1, "ENTER_NOW", lifecycle_state="ENTRY_TRIGGERED"),
        setup("AR", 2, "ALMOST_READY"),
        setup("EE", 3, "EARLY_ENTRY", lifecycle_state="EARLY_ENTRY_BUILDING"),
        setup("ET", 4, "EARLY_ENTRY", lifecycle_state="EARLY_TOUCH"),
        setup("RT", 5, "ALMOST_READY", lifecycle_state="WAITING_FOR_RETEST"),
        setup("MISS", 6, "EARLY_ENTRY", lifecycle_state="MISSED_ENTRY"),
        setup("INV", 7, "EARLY_ENTRY", lifecycle_state="INVALIDATED"),
        setup("EXP", 8, "EARLY_ENTRY", lifecycle_state="EXPIRED"),
        setup("WAIT", 9, "WAITING", tier="REVIEW"),
    ]
    original_rankings = [copy.deepcopy(row["ranking"]) for row in rows]

    assert scanner.stock_mission_workflow_bucket(rows[0]) == "TRADE_NOW"
    assert scanner.stock_mission_workflow_bucket(rows[1]) == "WATCH_CLOSELY"
    assert scanner.stock_mission_workflow_bucket(rows[2]) == "WATCH_CLOSELY"
    assert scanner.stock_mission_workflow_bucket(rows[3]) == "WATCH_CLOSELY"
    assert scanner.stock_mission_workflow_bucket(rows[4]) == "WATCH_CLOSELY"
    assert scanner.stock_mission_workflow_bucket(rows[5]) == "RESOLVED"
    assert scanner.stock_mission_workflow_bucket(rows[6]) == "RESOLVED"
    assert scanner.stock_mission_workflow_bucket(rows[7]) == "RESOLVED"
    assert scanner.stock_mission_workflow_bucket(rows[8]) == "WATCH_CLOSELY"

    snapshot = scanner.stock_mission_workflow_snapshot(rows, [], universe="test", update_movements=False)
    assert snapshot["enabled"] is False
    assert snapshot["trade_now_count"] == 1
    assert snapshot["watch_closely_count"] == 5
    assert snapshot["resolved_count"] == 3
    assert [row["ranking"] for row in rows] == original_rankings


def test_movement_diagnostics_are_generation_aware():
    scanner._stock_mission_workflow_previous.pop("mission-test", None)
    watch = setup("MOVE", 1, "ALMOST_READY", lifecycle_state="EARLY_ENTRY_BUILDING")
    first = scanner.stock_mission_workflow_snapshot([watch], [], universe="mission-test", update_movements=True)
    assert first["movements"]["entered_watch_closely"] == 1

    trade_now = setup("MOVE", 1, "ENTER_NOW", lifecycle_state="ENTRY_TRIGGERED")
    second = scanner.stock_mission_workflow_snapshot([trade_now], [], universe="mission-test", update_movements=True)
    assert second["movements"]["watch_closely_to_trade_now"] == 1

    replaced = setup("MOVE", 1, "EARLY_ENTRY", lifecycle_state="EARLY_ENTRY_BUILDING", entry=101.0)
    third = scanner.stock_mission_workflow_snapshot([replaced], [], universe="mission-test", update_movements=True)
    assert third["movements"]["entered_watch_closely"] == 1
    assert third["movements"]["plan_replacements_or_removed"] == 1


if __name__ == "__main__":
    test_bucket_classification_and_strategy_fields_are_read_only()
    test_movement_diagnostics_are_generation_aware()
    print("mission_workflow_v1.py passed")
